### PR TITLE
feat: Support for login and registration via a browser custom tab

### PR DESCRIPTION
**Pull request by xitij2000**
This change adds support for logging in and registering a new account using the
browser. This can be useful for cases where the only way to log into the
instatance is via a custom third-party auth provider.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @xitij2000!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/openedx-mobile-maintainers`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Review by navinkarkera — APPROVED** [review]
@xitij2000 Nice work! :+1: 

:+1:
- [x] I tested this: (tested login in both real hardware as well as emulator)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

**Comment by mphilbrick211** [comment]
@openedx/openedx-mobile-maintainers - is someone able to please take a look at this?

**Comment by volodymyr-chekyrta** [comment]
> @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?

I'll review it 👀 
@xitij2000, could you change the target branch to develop and fix conflicts if they appear?
Thank you!

**Comment by xitij2000** [comment]
> > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> 
> I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!

@volodymyr-chekyrta  I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that. 

Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?

**Comment by volodymyr-chekyrta** [comment]
> > > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> > 
> > 
> > I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!
> 
> @volodymyr-chekyrta I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that.
> 
> Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?

@xitij2000 I checked the approach, and it is good 👍; I have a few comments according to the code, but it will be better to address them after resolving conflicts

**Comment by xitij2000** [comment]
> > > > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> > > 
> > > 
> > > I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!
> > 
> > 
> > @volodymyr-chekyrta I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that.
> > Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?
> 
> @xitij2000 I checked the approach, and it is good 👍; I have a few comments according to the code, but it will be better to address them after resolving conflicts

I've rebased the code, but still need to test that everything works.

File: app/src/main/java/org/openedx/app/AppActivity.kt
Comment (volodymyr-chekyrta) [inline app/src/main/java/org/openedx/app/AppActivity.kt:161]: Please use `SignInFragment.newIntance(...)` method

File: auth/src/main/java/org/openedx/auth/data/repository/AuthRepository.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/data/repository/AuthRepository.kt:51]: `oauth2Callback` to constants

File: auth/src/main/java/org/openedx/auth/presentation/logistration/LogistrationFragment.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/logistration/LogistrationFragment.kt:83]: Please move `/register` to constants

File: auth/src/main/java/org/openedx/auth/presentation/signin/SignInFragment.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/signin/SignInFragment.kt:48]: Please move arguments like `"auth_code"` to companion object constants.
e.g.
```
companion object {
        private const val ARG_COURSE_ID = "courseId"
        private const val ARG_INFO_TYPE = "info_type"
        private const val ARG_AUTH_CODE = "auth_code"

        fun newInstance(courseId: String?, infoType: String?): SignInFragment {
            val fragment = SignInFragment()
            fragment.arguments = bundleOf(
                ARG_COURSE_ID to courseId,
                ARG_INFO_TYPE to infoType
            )
            return fragment
        }
    }
```

File: auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt:193]: I think we can remove the debug log from here

File: auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt:306]: `if(state.isBrowserLoginEnabled)` -> `if (state.isBrowserLoginEnabled)`

File: auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt:308]:
```suggestion
                    } else {
                        keyboardController?.hide()
```

File: auth/src/main/java/org/openedx/auth/presentation/sso/BrowserAuthHelper.kt
Comment (volodymyr-chekyrta) [inline auth/src/main/java/org/openedx/auth/presentation/sso/BrowserAuthHelper.kt:21]: It probably will be good to move the params to constants as well

File: core/src/main/java/org/openedx/core/config/Config.kt
Comment (volodymyr-chekyrta) [inline core/src/main/java/org/openedx/core/config/Config.kt:36]: duplicates:
```
fun getAppId(): String {
    return getString(APPLICATION_ID, "")
}
```

File: docs/how-tos/auth-using-browser.rst
Comment (volodymyr-chekyrta) [inline docs/how-tos/auth-using-browser.rst:3]: Just wanted to point out that I like this doc 👍

**Review by volodymyr-chekyrta — CHANGES_REQUESTED** [review]
Minor changes, mostly move value to constant:

**Comment by xitij2000** [comment]
@volodymyr-chekyrta  I will allocate some time soon to go through these properly and update the PR. Thanks for the great feedback!

**Comment by xitij2000** [comment]
@volodymyr-chekyrta  I've updated the PR based on your feedback. Do have a look!

**Comment by volodymyr-chekyrta** [comment]
The code looks good to me.
@marcotuts, would you like to check this feature from a product perspective before merging?

File: default_config/dev/config.yaml
Comment (xitij2000) [reply default_config/dev/config.yaml:81]: Yes, of course! Sorry.

**Comment by marcotuts** [comment]
Would it be possible to clarify the usage for this, and if possible any screenshots of it, even if just from local dev? 

I want to make sure I understand the flow here, is the configuration such that you are either using the native login / register (on by default?) or usign the browser mechanism? 

I ask becasue we also merged in a secondary web login button that can be optionally configured on the native view to direct users to a browser based login (ex: for corporate logins or other options not available natively). 

I still think the ability to choose a full browser login is a nimportant addition to the app, so I dont consider the two described features above as conflicting but I want to make sure I understand the changes and what the level of theming might be, or how the browser window renders (overlaps, embedded, etc). 

Thanks @xitij2000 for this contribution!

**Comment by xitij2000** [comment]
@marcotuts Sure! It will be hard to do it through screenshots since it a flow, but I will try to make a screencap and post it here. 

In this particular implementation, the web-based login repalces the native login. So when you click on the login button, it opens a custom browser tab, which is essentially a full-screen browser window (AFAIK it will use your default browser) where the OAuth2 based login flow will be initiated. When the user completes the log in in this browser windows (it could be using a username and password or whatever third-party auth system the LMS is configured to use) it will redirect back to the app with the code and the app will fetch a login token from that. 

From what I understand the feature you're describing adds the above login flow as an option during runtime rather than build time (which is what this PR does). I think if that is the case, we could simplify this PR to use that flow. Is my understanding correct?

**Comment by xitij2000** [comment]
@marcotuts  Here is a video of this at work:
[Screen_recording_20241209_172251.webm](https://github.com/user-attachments/assets/ce988634-5dcd-42c2-bd11-3ac1957c9a81)

**Comment by xitij2000** [comment]
> Thanks for the video! This looks great, thanks again for the contribution. It might be worth eventually exploring the embedded mode for this view but it seamlessly loads and returns to the app in the demo so that's 99% of what we need :D
> 
> product 👍


I am not very knowledgeable about Android so I followed the advice I heard about WebViews for OAuth2 which instead [recommended using custom tabs](https://support.google.com/faqs/answer/12284343?hl=en-GB).

**Review by volodymyr-chekyrta — APPROVED** [review]

**Pull request merged**
